spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_structural.py

```diff
@@ -0,0 +1,514 @@
+        self._lat_distance_probabilities = \
+            self.generate_distance_probability_array(
+                self._p_form_lateral,
+                self._sigma_form_lateral)
+
+    @overrides(AbstractSynapseDynamicsStructural.get_parameter_names)
+    def get_parameter_names(self):
+        # TODO
+        return []
+
+    def distance(self, x0, x1, grid=np.asarray([16, 16]),
+                 type='euclidian'):  # @ReservedAssignment
+        x0 = np.asarray(x0)
+        x1 = np.asarray(x1)
+        delta = np.abs(x0 - x1)
+        #     delta = np.where(delta > grid * .5, delta - grid, delta)
+        #     print delta, grid
+        if delta[0] > grid[0] * .5 and grid[0] > 0:
+            delta[0] -= grid[0]
+
+        if delta[1] > grid[1] * .5 and grid[1] > 0:
+            delta[1] -= grid[1]
+
+        if type == 'manhattan':
+            return np.abs(delta).sum(axis=-1)
+        return np.sqrt((delta ** 2).sum(axis=-1))
+
+    def generate_distance_probability_array(self, probability, sigma):
+        euclidian_distances = np.ones(self._grid ** 2) * np.nan
+        for row in range(euclidian_distances.shape[0]):
+            for column in range(euclidian_distances.shape[1]):
+                if self._grid[0] > 1:
+                    pre = (row // self._grid[0], row % self._grid[1])
+                    post = (column // self._grid[0], column % self._grid[1])
+                else:
+                    pre = (0, row % self._grid[1])
+                    post = (0, column % self._grid[1])
+                euclidian_distances[row, column] = self.distance(
+                    pre,
+                    post,
+                    grid=self._grid,
+                    type='euclidian')
+        largest_squared_distance = np.max(euclidian_distances ** 2)
+        squared_distances = np.arange(largest_squared_distance)
+        raw_probabilities = probability * (
+            np.exp(-(squared_distances) / (2 * (sigma ** 2))))
+        quantised_probabilities = raw_probabilities * ((2 ** 16) - 1)
+        # Quantize probabilities and cast as uint16 / short
+        unfiltered_probabilities = quantised_probabilities.astype(
+            dtype="uint16")
+        # Only return probabilities which are non-zero
+        filtered_probabilities = unfiltered_probabilities[
+            unfiltered_probabilities > 0]
+        if filtered_probabilities.size % 2 != 0:
+            filtered_probabilities = np.concatenate(
+                (filtered_probabilities,
+                 np.zeros(
+                     filtered_probabilities.size % 2, dtype="uint16")))
+
+        return filtered_probabilities
+
+    @property
+    def p_rew(self):
+        return self._p_rew
+
+    @property
+    def actual_sdram_usage(self):
+        return self._actual_sdram_usage
+
+    @property
+    def approximate_sdram_usage(self):
+        return self._approximate_sdram_usage
+
+    def write_parameters(self, spec, region, machine_time_step, weight_scales,
```

Comment:
This is a very large method. Breaking it up into a few pieces with distinct roles would make it much easier to understand.

Reply:
Done